Source/Data/Remotes/ParleyRemote.swift

```diff
@@ -4,9 +4,60 @@ import ObjectMapper
 import UIKit
 import Foundation
 
+
+public typealias ParleyRemoteInterceptor = ParleyRemoteRequestInterceptorProtocol & ParleyRemoteResponseInterceptorProtocol
+
+public protocol ParleyRemoteRequestInterceptorProtocol: AnyObject {
+    func httpRequestWillStart(with request: inout URLRequest) throws
+}
+
+public protocol ParleyRemoteResponseInterceptorProtocol: AnyObject {
+    func httpResponseReceived(response: HTTPURLResponse) throws
+}
+
+internal class ParleyRemoteInterceptorProxy: RequestInterceptor {
+
+    private let interceptors: [ParleyRemoteInterceptor]
+    
+    init(interceptors: [ParleyRemoteInterceptor]) {
+        self.interceptors = interceptors
+    }
+    
+    func adapt(_ urlRequest: URLRequest, for session: Session, completion: @escaping (Result<URLRequest, Error>) -> Void) {
+        guard interceptors.count > 0 else {
+            completion(.success(urlRequest))
+            return
+        }
+        var editableUrlRequest = urlRequest
+        
+        interceptors.forEach {
+            do {
+                try $0.httpRequestWillStart(with: &editableUrlRequest)
+            } catch {
+                completion(.failure(error))
+            }
+        }
+        completion(.success(editableUrlRequest))
```

Comment:
There's a possibility this code crashes. Make sure completion handlers always are called max once. This for loop might cause them to fire multiple times (when there's multiple interceptors added). Depending on the earlier discussion about throwing the request, the fix for this differs.

Reply:
Good suggestion.